### DISCUSSION TITLE
How to use Spring to extract complex structural data

### DISCUSSION THREAD

**Discussion by muchengyang** [spring-projects/spring-ai#1122] [Q&A]
Seeking advice from experts, my requirement is to have the large model extract data according to the given JSON schema. I have tried many solutions and consulted a lot of information, but have not been able to achieve the desired results. I hope that experienced experts can share their experience with me
Here is my example
Prompt：
Your response should be in JSON format.
Do not include any explanations, only provide a RFC8259 compliant JSON response following this format without deviation.
Do not include markdown code blocks in your response.
Remove the ```json markdown from the output.
Here is the JSON Schema instance your output must adhere to:
```{
    "$schema": " https://json-schema.org/draft/2020-12/schema ",
    "type": "object",
    "properties": {
        "WardName": {
            "type": "string"
        },
        "Bednumber": {
            "type": "string"
        },
        "Bodytemperature": {
            "type": "number"
        },
        "Pulse": {
            "type": "number"
        },
        "Systolicpressure": {
            "type": "number"
        },
        "Date": {
            "type": "string",
            "format": "date"
        },
        "Recordtime": {
            "type": "string",
            "format": "time"
        },
        "Diastolicbloodpressure": {
            "type": "number"
        }
    }
}`` 
1. 318 beds in the rehabilitation ward, pulse 80, systolic blood pressure 120, diastolic blood pressure 80, body temperature 37, tomorrow morning at 10am
2. Rehabilitation ward 318 beds, pulse rate 80, tomorrow morning at 10am
3. 318 beds
I hope that the inputs 1, 2, and 3 mentioned above can be accurately JSON. If no value is extracted, the assignment will be empty. Random assignment of large models is not allowed, and time can be calculated correctly

**Comment by ThomasVitale** [comment]
Do you have a Java object corresponding to that JSON schema, or do you need to pass the schema explicitly?
Spring AI can generate a valid standard JSON schema from a Java object and enhance the prompt for you to instruct the model on how to structure the response.

I have an example here for the "structured data extraction" use case, in case it might help: https://github.com/ThomasVitale/llm-apps-java-spring-ai/blob/main/00-use-cases/structured-data-extraction/src/main/java/com/thomasvitale/ai/spring/StructuredDataExtractionService.java Incidentally, it's also healthcare related. You can find info about how to run the app locally [here](https://github.com/ThomasVitale/llm-apps-java-spring-ai/blob/main/00-use-cases/structured-data-extraction/README.md). Once the app is running, you can call it as follows:

`http --raw "I'm visiting Jon Snow. The blood pressure looks fine: 120/80. The temperature is 36 degrees. The diagnosis is: he knows nothing." :8080/extract`

The model will extract the data into a structured patient journal object.

This is the prompt I used, which is then enhanced by Spring AI with instructions and info about the JSON schema for the object used as the return type.

```
Extract structured data from the provided text.
If you do not know the value of a field asked to extract,
do not include any value for the field in the result.

---------------------
TEXT:
{text}
---------------------
```

A couple of things to consider for this use case:

- the temperature should be set to 0 to improve the accuracy of the result
- the schema shouldn't have required fields or else the model would be forced to make things up in order to produce a compliant result, which is not what you want. The schema you shared is fine since each field is optional.

**Comment by muchengyang** [comment]
@ThomasVitale 
Thank you very much for your answer. Today, I will try my own business using the method you provided. If I encounter any questions that I don't understand, I hope you can help me solve them again

**Reply by muchengyang** [reply]
In my scenario, need to pass the schema explicitly， Because the structure that needs to be extracted from the text is not fixed every time

**Comment by muchengyang** [comment]
The model is unable to return correct results when dealing with time-related queries. According to my research, the model lacks a reference date. I've tried providing today's date as a reference in the prompt, but it still returns incorrect values. Do you have any good solutions for this?
```java
	private static String getFormat(String jsonSchema) {
		String template =
				"Extract structured data from the provided text.\n"
						+ "Your response should be in JSON format.\n"
						+ "If you do not know the value of a field asked to extract,\n"
						+ "do not include any value for the field in the result.\n"
						+ "If you cannot extract the value of any field, return an empty JSON directly.\n"
						+ "Do not include any explanations, only provide a RFC8259 compliant JSON response following this format without deviation.\n"
						+ "Do not include markdown code blocks in your response.\n"
						+ "Remove the ```json markdown from the output.\n"
						+ "Here is the JSON Schema instance your output must adhere to:\n"
						+ "```%s```\n";
		return String.format(template, jsonSchema);
	}
	@PostMapping("/chat")
	public String chat(String content){
		SystemMessage systemMessage = new SystemMessage(getFormat("{\"$schema\":\"https://json-schema.org/draft/2020-12/schema \",\"type\":\"object\",\"properties\":{\"bck03\":{\"type\":\"string\",\"description\":\"Ward Name\"},\"bcq04b\":{\"type\":\"number\",\"description\":\"Bed number\"},\"diastolicPressValue\":{\"type\":\"number\",\"description\":\"Diastolic blood pressure\"},\"pulseValue\":{\"type\":\"number\",\"description\":\"Pulse\"},\"recordDate\":{\"type\":\"string\",\"description\":\"Date\",\"format\":\"yyyy-MM-dd\"},\"recordTime\":{\"type\":\"string\",\"description\":\"Time\",\"format\":\"hh:mm:ss\"},\"systolicPressValue\":{\"type\":\"number\",\"description\":\"Systolic pressure\"},\"tempValue\":{\"type\":\"number\",\"description\":\"Body temperature\"}}}\n"));
		return  ChatClient.create(chatModel).prompt()
				.messages(systemMessage)
				.user(u -> u.text("""
												---------------------
								                TEXT:
								                {text}
								                ---------------------
								""").param("text", content))
				.call()
				.content();
	}
	```
